feat: Add `esquery` selector textfield & highlighting of matched code

pkerschbaum opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [OpenJS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

-   [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request?

MVP for `esquery` selector playground (#79): textfield to enter selectors, matching code get's highlighted:

![image](https://github.com/user-attachments/assets/ce48853b-296b-4341-a28e-de298d699679)

Future ideas:

* add matches count, tooltips for matches, or syntax highlighting of the selector itself like https://regexr.com/

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

* add reusable component `TextField`, styled similar to `Button`
* add textfield for esquery selector in first (left) panel
* add switch to nav bar to toggle the esquery selector textfield
  * is by default off --> people need to opt-into displaying the esquery selector textfield
* add Codemirror extension + CSS to highlight text ranges via its "Decoration" feature
* if the esquery toggle is enabled, and the esquery selector matches nodes
  * highlight text ranges in the editor
  * highlight nodes in the AST tree tool

#### Related Issues

Fixes #79

#### Is there anything you'd like reviewers to focus on?

<!-- markdownlint-disable-file MD004 -->

nzakas commented:
Thanks for putting this together. Some notes on the UI:

1. I don't think having a textbox span the entire width of the app is the right way to go. I was thinking it should be inside one of the panes. 
1. If we include ESQuery, it should work for all languages, not just JavaScript, so I don't think toggling in the language dropdown is the right location. Instead, what about adding a button that is next to the theme button in the toolbar?
2. Highlighting the text in the editor is great. I think we also need to highlight the nodes in the tree from the start.

For some perspective, check out the typescript-eslint playground that has an ESQuery filter:
https://typescript-eslint.io/play/

pkerschbaum commented:
sounds good, will do!

pkerschbaum commented:
I implemented all suggestions:

* textbox is now in the left panel
* esquery selector applies to all languages (didn't know that is possible, had to read through https://eslint.org/blog/2024/07/whats-coming-next-for-eslint/ etc ^^)
* the nodes in the tree view (accordions) are also highlighted now (border: width increase + primary color)

In order to have 1 place to apply the match nodes via the esquery selector, I added a `useAST` hook where the JS/CSS/Markdown/JSON code-to-AST parsing happens now.

The esquery matched nodes need to be prop-drilled everywhere, and the accordion highlighting is duplicated. That's not ideal but I think in the future we could refactor the code anyways to get the accordion code generalized at a central place.

pkerschbaum commented:
@nzakas @amareshsm plz re-review :)

amareshsm commented:
Looks great! In addition to Nzaka's feedback, I’ve listed a couple of minor points below.

1. When I type a program in the ESQuery Selector, the code gets highlighted, but the program node isn't highlighted in the AST preview.

![image](https://github.com/user-attachments/assets/a07d715a-3a8d-4777-892e-0477f3d0b549)

2. Let's add a placeholder text to the ESQuery Selector text field.

pkerschbaum commented:
thx for review!

1. I removed the toggle.
2. I fixed the issue that the root node of JS AST tree view would not get highlighted (forgot 1 place to put the logic - we really need to unify some logic ^^)
3. That dark focus outline is currently aligned with how the focus looks like for all other buttons - to be consistent. Still, do you want me to change it to purple shade?  
   <img width="293" alt="image" src="https://github.com/user-attachments/assets/7f05bbd5-d772-413e-9905-e472bf9b5d37" />
5. I added the red background if the selector does not match anything; as `destructive` of the available tailwind colors was too harsh, see screenshot...  
  <img width="960" alt="image" src="https://github.com/user-attachments/assets/5b47ba44-e9ad-464d-92a6-a9027e6a2978" /> ...so I added one of the `codemirror-themes.css` as `nonmatchingEsquerySelector` tailwind color:  
   <img width="964" alt="image" src="https://github.com/user-attachments/assets/83d21008-5a10-4502-beb6-971047291156" />
7. I could not reproduce the crash - but I made the access onto the new state property more safe, hope that avoids it.

pkerschbaum commented:
Also added the `placeholder`

@nzakas @amareshsm plz re-review :)

amareshsm commented:
> That dark focus outline is currently aligned with how the focus looks like for all other buttons - to be consistent. Still, do you want me to change it to purple shade?

+1 with updating the color to one of the purple shades to align with the design we use across other ESLint sites. We’re using the color shade --color-primary-800 in other parts of the site.

nzakas commented:
@pkerschbaum no need to ask for reviews, we get the notifications. :smile: We just have a lot to get through every day so your patience is appreciated.

Ah, I hadn't even noticed that the focus outline was the same in other form controls. We can leave it as-is for now. No need to bundle unrelated changes in this PR.

Let's be sure to stick to the colors that are already in this file:
https://github.com/eslint/code-explorer/blob/main/src/codemirror-themes.css

These are all of our theme colors and we don't want to stray outside of that. I think one of the `color-rose` values would work for the textbox (though we may need a different color for dark mode).

Overall this is looking really good.

pkerschbaum commented:
I implemented the rose-red highlighting when the esquery selector has no matches.  
I kept the black focus outline - I suggest to change that in a followup PR for all inputs together 👍  
So, should be ready for re-review

amareshsm commented:
In JSON language, when I enter "STRING" in the ESQuery selector, it causes the UI to break.

pkerschbaum commented:
> In JSON language, when I enter "STRING" in the ESQuery selector, it causes the UI to break.

damn, didn't see that... fixed with 4cd9966054c921066262327cec94626bb81465d8

nzakas reviewed (commented):
This is looking so good!

There's one strange behavior I found: When the ESQuery selector has no matches, the background turns reddish but only if the cursor isn't over the textbox. If the cursor is over the textbox, the background turns a lighter color and the reddish color is lost.

I think the reddish color should stay regardless of mouse position. What do you think?

pkerschbaum commented:
> This is looking so good!
> 
> There's one strange behavior I found: When the ESQuery selector has no matches, the background turns reddish but only if the cursor isn't over the textbox. If the cursor is over the textbox, the background turns a lighter color and the reddish color is lost.
> 
> I think the reddish color should stay regardless of mouse position. What do you think?

I agree that it feels strange - I changed it so that the reddish background color stays also on hover.

nzakas reviewed (approved):
LGTM. Awesome work on this.

Would like @amareshsm to approve before merging.

harish-sethuraman commented on src/App.tsx:
```diff
@@ -22,8 +28,16 @@ function App() {
+									<EsquerySelectorInput />
 									<Editor
 										value={code[language]}
+										highlightedRanges={
+											!astParseResult.ok
+												? undefined
+												: convertNodesToRanges(
+														astParseResult.esqueryMatchedNodes,
+													)
```
can we switch the case? and remove the `!`?
```
astParseResult.ok ?  convertNodesToRanges(astParseResult.esqueryMatchedNodes) : undefined
```

pkerschbaum replied:
done

harish-sethuraman commented:
for some reason when I select a nested element the right accordion doesn't highlight? is this expected? 

selector used: `ImportDeclaration > Literal`
<img width="1726" alt="Screenshot 2025-02-25 at 11 27 02 AM" src="https://github.com/user-attachments/assets/4ad1eb5b-3529-45f1-a112-e74512ea5b95" />

pkerschbaum commented:
> ImportDeclaration > Literal

The accordion is highlighted:

![image](https://github.com/user-attachments/assets/65af13bc-5d4b-4dc8-a376-ffb362c93345)

nzakas commented:
@harish-sethuraman the right side doesn't automatically open to show the highlighted nodes, but they are higlighted.

I think we need https://github.com/eslint/code-explorer/issues/64 on top of this to really get to where we'd like to be.

amareshsm reviewed (approved):
LGTM. Thanks!
